components/markets/SellFullSetModal.tsx

```diff
@@ -1,139 +1,179 @@
+    wallets.getActiveSigner(),
+    pool,
+  );
+
+  const [transacting, setTransacting] = useState(false);
+  const [amount, setAmount] = useState<string>("0");
+  const [maxTokenSet, setMaxTokenSet] = useState<Decimal>(new Decimal(0));
 
-    const handleAmountChange = (amount: string) => {
-      setAmount(amount);
-    };
+  useEffect(() => {
+    let lowestTokenAmount: Decimal = null;
+    balances?.forEach((balance) => {
+      const free = new Decimal(balance.free.toNumber());
+      if (!lowestTokenAmount || free.lessThan(lowestTokenAmount)) {
+        lowestTokenAmount = free;
+      }
+    });
+    setMaxTokenSet(lowestTokenAmount ?? new Decimal(0));
+  }, [balances]);
 
-    const handleSignTransaction = async () => {
-      if (Number(amount) > maxTokenSet.toNumber() || Number(amount) === 0)
-        return;
+  const handleAmountChange = (amount: string) => {
+    setAmount(amount);
+  };
+
+  const handleSignTransaction = async () => {
+    if (
+      Number(amount) > wallets.activeBalance.toNumber() ||
+      Number(amount) === 0
+    ) {
+      return;
+    }
+
+    setTransacting(true);
+
+    if (hasMarketMethods<Context>(market)) {
       const signer = wallets.getActiveSigner();
 
-      await marketStore.market.sellCompleteSet(
-        signer,
-        new Decimal(amount).mul(ZTG).toNumber(),
-        extrinsicCallback({
-          notificationStore,
-          successCallback: () => {
-            notificationStore.pushNotification(
-              `Sold ${new Decimal(amount).toFixed(1)} full sets`,
-              { type: "Success" },
-            );
-            modalStore.closeModal();
-            marketsStore.getMarket(marketStore.id);
-          },
-          failCallback: ({ index, error }) => {
-            notificationStore.pushNotification(
-              store.getTransactionError(index, error),
-              {
-                type: "Error",
-              },
-            );
+      notificationStore?.pushNotification("Transacting...", {
+        autoRemove: true,
+      });
+
+      const result = await market
+        .sellCompleteSet({
+          amount: new Decimal(amount).mul(ZTG).toNumber(),
+          signer,
+          hooks: {
+            inBlock: () => {
+              notificationStore.pushNotification(
+                `In block: sold ${new Decimal(amount).toFixed(
+                  1,
+                )} full sets. Waiting for finalization..`,
+                { type: "Info", autoRemove: true, lifetime: 45 },
+              );
+              setTransacting(false);
+              modalStore.closeModal();
+            },
+            retracted: () => {
+              notificationStore.pushNotification(
+                `Transaction retracted and will take longer than usual to be finalized.`,
+                { type: "Info", lifetime: 45 },
+              );
+            },
           },
-        }),
-      );
-    };
+        })
+        .asEither();
+
+      if (isRight(result)) {
```

Comment:
Some feedback on the sdk. I'm a bit confused by these util functions (`isRight`, `asEither`, `unleft`, etc). They are documented but often refer to other functions in the SDK. For example `unleft` requires and understanding of `right` and `unwrapOr`. Though I don't really understand the concept of left and right in the sdk, is this documented somewhere?

Reply:
I was wondering the same thing.